lib/active_merchant/billing/gateways/aafes.rb

```diff
@@ -0,0 +1,172 @@
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript
+      end
+
+      private
+
+      def add_invoice(post, money, options)
+        post[:amount] = amount(money)
+        post[:currency] = (options[:currency] || currency(money))
+      end
+
+      def parse(body)
+        xml = Nokogiri::XML(body)
+        Hash.from_xml(xml.to_s)['Message']
+      end
+
+      def commit(request)
+        url = (test? ? test_url : live_url)
+        response =
+          begin
+            parse(ssl_post(url, request, headers))
+          rescue StandardError => error
+            parse(error.response.body)
+          end
+
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def headers
+        {
+          'Content-Type' => 'application/xml'
+        }
+      end
+
+      def success_from(response)
+        return true if response.dig('Response', 'ReasonCode').to_i.between?(0, 100)
+
+        return false
+      end
+
+      def message_from(response)
+        response.dig('Response', 'ResponseType')
+      end
+
+      def authorization_from(response) end
+
+      def build_xml_request
+        builder = Nokogiri::XML::Builder.new(encoding: 'UTF-8') do |xml|
+          xml['cm'].Message(
+            'TypeCode' => 'Request',
+            'MajorVersion' => '3',
+            'MinorVersion' => '4',
+            'FixVersion' => '0',
+            'xmlns:cm' => 'http://www.aafes.com/credit') do
+            yield(xml)
+          end
+        end
+
+        builder.to_xml
+      end
+
+      def error_code_from(response)
+        unless success_from(response)
+          # TODO: lookup error code for this response
+        end
+      end
+
+      def add_headers(xml, options)
+        xml['cm'].Header do
+          xml['cm'].IdentityUUID(@identity_uuid)
+          xml['cm'].LocalDateTime(Time.now.utc.iso8601)
+          xml['cm'].SettleIndicator(false)
+          xml['cm'].OrderNumber(options[:order_id])
+          xml['cm'].transactionId(options[:transaction_id])
+          xml['cm'].termId(options[:term_id])
+          xml['cm'].Comment(options[:comment])
+          xml['cm'].CustomerID(options[:customer_id])
+        end
+      end
+
+      def add_milstar_auth(xml, amount, payment, options)
+        xml['cm'].Request('RRN' => options[:rrn]) do
+          xml['cm'].Media('Milstar')
+          xml['cm'].RequestType('Sale')
+          xml['cm'].InputType('Keyed')
+          xml['cm'].Token('Token')
+          xml['cm'].Account(payment.payment_data)
+          xml['cm'].Expiration(payment.metadata[:expiration])
```

Comment:
I'm not entirely sure where this expiration date information is coming from, but would it be appropriate to follow [the way that Ruthan implemented adding expiration dates to third-party vaulted payment methods for Litle](https://github.com/activemerchant/active_merchant/pull/3606/files)?

Reply:
The expiration date will be coming from Priceline, along with the Milstar token and billing zip. And Priceline gets the expiration from AAFES in the format of `YYMM`. I don't think we need to do any tinkering with it, since AAFES gives it to us in the format they expect it.